radicle-node/src/tests/e2e.rs

```diff
@@ -379,3 +383,78 @@ fn test_connection_crossing() {
 
     assert!(s1 ^ s2, "Exactly one session should be established");
 }
+
+//
+// Alice -- Bob
+//
+// Bob clones from Alice's repo without tracking Alice first. This is OK
+// as we always fetch `Namespaces::All` for a new repo.
+//
+// Alice will not fetch `RefsAnnouncement` from Bob as Alice does not
+// track Bob as `Trusted`.
+#[test]
+fn test_trusted_node_negative() {
+    logger::init(log::Level::Debug);
+
+    let mut environment = Environment::new();
+    let alice = environment.node("alice");
+    let bob = environment.node("bob");
+    let working = environment.tmp().join("working");
+    let rid = Id::from_str("z42hL2jL4XNk6K8oHQaSWfMgCL7ji").unwrap();
+
+    let mut alice = alice.spawn(Config::default());
+    let mut bob = bob.spawn(Config::default());
+
+    alice.connect(&bob);
+    bob.connect(&alice);
+
+    alice.routes_to(&[]);
+    bob.routes_to(&[]);
+
+    // Initialize a repo as Alice.
+    fixtures::repository(working.join("alice"));
+    alice
+        .rad(
+            "init",
+            &[
+                "--name",
+                "heartwood",
+                "--description",
+                "Radicle Heartwood Protocol & Stack",
+                "--default-branch",
+                "master",
+                "--announce",
+            ],
+            working.join("alice"),
+        )
+        .unwrap();
+
+    alice.routes_to(&[(rid, alice.id)]);
+    bob.routes_to(&[(rid, alice.id)]);
+
+    // Verify that Bob can clone from Alice.
+    bob.rad("clone", &[rid.to_string().as_str()], working.join("bob"))
+        .unwrap();
+
+    // Enough time for Bob to be able to fetch.
+    thread::sleep(time::Duration::from_secs(1));
+
+    // Routes are updated.
+    alice.routes_to(&[(rid, alice.id), (rid, bob.id)]);
+    bob.routes_to(&[(rid, alice.id), (rid, bob.id)]);
+
+    // Bob has Alice's refs.
+    bob.storage
+        .repository(rid)
+        .unwrap()
+        .remote(&alice.id)
+        .unwrap();
+
+    // Alice will not fetch Bob's refs as Alice does not track Bob.
+    assert!(alice
+        .storage
+        .repository(rid)
+        .unwrap()
+        .remote(&bob.id)
+        .is_err());
+}
```

Comment:
I think this test is testing some of the stuff I was suggesting for the other test. I would suggest that we just have the other test, as the e2e tests are very slow and only really useful if we want to test the networking protocols. A service logic test should be enough since we only modified the service.

Reply:
I don't have accurate measurements but this e2e test runs pretty fast on my not-so-fast laptop.  The reason I am keeping this test is that it is much more easier to trigger the path I wanted:  Bob announces its refs after everyone already has the repo. (I was struggling to write a no-super-complicated test case to cover such path in the service `tests.rs`) .

Reply:
Hmm I don't see why that would be complicated. You would just call `alice.receive(bob_refs)` and `eve.receive(bob_refs)`.

The service tests are preferred for logic tests as:
* There are no false negatives (e2e tests can fail for a number of reasons)
* They run in 1/100 of the time
* It's much clearer to see what is actually being tested vs. what is just a side-effect of setting up the test

Reply:
Suppose `alice` is the first node that creates the repo, how will `bob` create its own refs? One way I know is `clone`, which is done easily in `e2e` test: `NodeHandle.rad("clone", ..)`.  Is there a different way to do `clone` in the service test?

Reply:
What I'm suggesting in the other comment is that Bob doesn't need to actually have the refs. We can forge a message from Bob, with whatever refs we want, and have Bob sign it, without Bob actually having the inventory. This is usually done with `Peer::refs_announcement(rid)`. If for some reason that function doesn't do what you want it to do, I'd like to understand better, and we'd want to manually create the refs you need for the test.

Reply:
Sorry about all the confusions. Like I mentioned in zulip,  my intention is to test this path:  

Both peer A and peer B has repo R and their own refs.  Peer A does not fetch B's refs because A does not track B yet.  Later, peer A would track peer B and able to fetch B's refs.

So far, I can only came up with the test I've added in `e2e.rs`, which is pretty straightforward after today's updates, and it covers the scenarios above.  I agree that this test is not super fast, although it seems to be ok on my laptop.

I tried to write a different & easier / faster test in `tests.rs`, and so far not able to.  (I avoided to make the test super complicated as that will defeat the original purpose.)  But again I would be wrong playing with the test infra.  I will update later either here or separately if I learn to write a better test for this case.